utilities/scripts/python/diff.py

```diff
@@ -0,0 +1,156 @@
+    add_or_remove_flag: bool
+        A flag indicating whether the entry should be removed (False) or added (True).
+    """
+
+    val: any
+    add_or_remove_flag: bool
+
+
+@dataclass
+class Diff:
+    """
+    Represents the difference between two dictionaries.
+    """
+    diff: dict
+
+    def __init__(self, expected: dict, provided: dict, strict: bool):
+        if strict:
+            self.diff = _strict_diff_(expected, provided)
+        else:
+            self.diff = _inclusion_diff_(expected, provided)
+
+    def has_diff(self) -> bool:
+        return True if self.diff else False
+
+    def to_ascii_colored_string(
+        self,
+        obj_name_to_add: str,
+        obj_name_to_remove: str,
+    ) -> str:
+        """
+        Generates an ascii colored string representation of the diff.
+
+        Args:
+            obj_name_to_add (str): The name of the object to add.
+            obj_name_to_remove (str): The name of the object to remove.
+        """
+
+        def _impl_(
+            diff: dict,
+            obj_name_to_add: str,
+            obj_name_to_remove: str,
+            ident: str = "",
+            path: str = "",
+        ) -> str:
+            result = ""
+            if isinstance(diff, DiffEntry):
+                color = "green" if diff.add_or_remove_flag else "red"
+                minus_or_plus = "+" if diff.add_or_remove_flag else "-"
+                obj_name = (
+                    obj_name_to_add if diff.add_or_remove_flag else obj_name_to_remove
+                )
+
+                result += "\n------\n"
+                result += _add_color_(obj_name, "yellow")
+                result += f"{path}\n"
+                result += _add_color_(f"{minus_or_plus}{ident} {diff.val}", color)
+
+            if isinstance(diff, dict):
+                for key in diff:
+                    result += _impl_(
+                        diff[key],
+                        obj_name_to_add,
+                        obj_name_to_remove,
+                        ident + " ",
+                        path + "\n" + f"{ident}  {key}",
+                    )
+
+            if isinstance(diff, list):
+                for val in diff:
+                    result += _impl_(
+                        val, obj_name_to_add, obj_name_to_remove, ident, path
+                    )
+
+            return result
+
+        return _impl_(self.diff, obj_name_to_add, obj_name_to_remove)
+
+
+def _inclusion_diff_(expected: dict, provided: dict) -> dict:
+    """
+    Calculate an inclusion diff between the expected and provided inputs in a recursive manner.
+
+    Args:
+        expected: The expected input.
+        provided: The provided input.
+
+    Returns:
+        A dictionary representing the difference between the expected and provided inputs.
+    """
+    diff = {}
+    if not isinstance(expected, dict):
+        if expected != provided:
+            return [DiffEntry(expected, True), DiffEntry(provided, False)]
```

Comment:
This is not the return type in the function definition.  Its an array, not a dict.

Reply:
For simplicity, if your going to return an array and also a dict, just always return an array of dicts, and the array has either 1 or 2 entries.
Even better is NOT use an array but use a tuple, and return a tuple with either 1 or two dicts.

Reply:
fixed